File: Erlang/0002.erl
%% Definition for singly-linked list.
%%
%% -record(list_node, {val = 0 :: integer(),
%%                     next = null :: 'null' | #list_node{}}).

-spec add_two_numbers(L1 :: #list_node{} | null, L2 :: #list_node{} | null) -> #list_node{} | null.
add_two_numbers(L1, L2) ->
    add_two_numbers(L1, L2, 0).

-spec add_two_numbers(L1 :: #list_node{} | null, L2 :: #list_node{} | null, Carry :: integer()) -> #list_node{} | null.
add_two_numbers(null, null, 0) ->
    null;
add_two_numbers(null, null, Carry) when Carry > 0 ->
    #list_node{val = Carry, next = null};
add_two_numbers(L1, null, Carry) ->
    add_two_numbers(L1, #list_node{val = 0, next = null}, Carry);
add_two_numbers(null, L2, Carry) ->
    add_two_numbers(#list_node{val = 0, next = null}, L2, Carry);
add_two_numbers(#list_node{val = Val1, next = Next1}, #list_node{val = Val2, next = Next2}, Carry) ->
    Sum = Val1 + Val2 + Carry,
    NewVal = Sum rem 10,
    NewCarry = Sum div 10,
    #list_node{val = NewVal, next = add_two_numbers(Next1, Next2, NewCarry)}.

File: Elixir/0002.exs
# Definition for singly-linked list.
#
# defmodule ListNode do
#   @type t :: %__MODULE__{
#           val: integer,
#           next: ListNode.t() | nil
#         }
#   defstruct val: 0, next: nil
# end

defmodule Solution do
  @spec add_two_numbers(l1 :: ListNode.t() | nil, l2 :: ListNode.t() | nil) :: ListNode.t() | nil
  def add_two_numbers(l1, l2) do
    add_two_numbers(l1, l2, 0)
  end

  defp add_two_numbers(nil, nil, 0), do: nil

  defp add_two_numbers(l1, l2, carry) do
    val1 = if l1, do: l1.val, else: 0
    val2 = if l2, do: l2.val, else: 0

    sum = val1 + val2 + carry
    new_carry = div(sum, 10)
    new_val = rem(sum, 10)

    %ListNode{
      val: new_val,
      next: add_two_numbers(
        (if l1, do: l1.next, else: nil),
        (if l2, do: l2.next, else: nil),
        new_carry
      )
    }
  end
end